core/trie/proof.go

```diff
@@ -438,27 +441,32 @@ func ensureMonotonicIncreasing(proofKeys [2]*Key, keys []*felt.Felt) error {
 
 // compressNode determines if the node needs compressed, and if so, the len needed to arrive at the next key
 func compressNode(idx int, proofNodes []ProofNode, hashF hashFunc) (int, uint8, error) {
-	parent := &proofNodes[idx]
+	parent := proofNodes[idx]
 
 	if idx == len(proofNodes)-1 {
-		if parent.Edge != nil {
+		if _, ok := parent.(*Edge); ok {
 			return 1, parent.Len(), nil
 		}
 		return 0, parent.Len(), nil
 	}
 
-	child := &proofNodes[idx+1]
+	child := proofNodes[idx+1]
 
-	switch {
-	case parent.Edge != nil && child.Binary != nil:
-		return 1, parent.Edge.Path.len, nil
-	case parent.Binary != nil && child.Edge != nil:
+	switch parent := parent.(type) {
+	case *Edge:
+		if _, ok := child.(*Binary); !ok {
+			break
+		}
+		return 1, parent.Len(), nil
+	case *Binary:
+		if _, ok := child.(*Edge); !ok {
```

Comment:
I think you can move checking is it binary or edge before switch like that:
```
_, isChildBinary := child.(*Binary)
isChildEdge = !isChildBinary
```
and use them of just one variable later in switch cases